### ISSUE TITLE
Support where clauses on trait declarations

### ISSUE THREAD

**Issue by kyouko-taiga** [hylo-lang/hylo#1112]
It should be possible to add a where clause on a trait declaration to constrain the associated types and values inherited by refinement. For example:
```
trait MovableCollection: Movable, Collection where Element: Movable { ... } 
```

**Comment by dabrahams** [comment]
+1, although unless we can think of a collection that _cannot_ be movable when its elements are, that particular example should probably be a trait alias of some kind…

**Comment by kyouko-taiga** [comment]
> unless we can think of a collection that cannot be movable when its elements are, that particular example should probably be a trait alias of some kind…

Good point. This particular example is not a good one. It came up because I've discovered that `MutableCollection` probably requires `Element: Movable`. That's because it seems having a mutable subscript is not sufficient to implement most generic in-place mutation algorithms. You also need `swap_at(_:_:)`.

**Comment by dabrahams** [comment]
And `swap_at` is not implementable in terms of… well, you can't copy… so… yeah.

It might be useful to think for just a moment first about whether there's anything we want to be able to do with a `MutableCollection` of immovable objects.  My biggest problem in this space is that it's hard to imagine what kinds of things should be immovable.  If such things exist, I can imagine making a fixed-sized buffer of them, where they all get constructed in-place.  I guess remote things are immutable, right?

So, I think the unfortunate truth is that most generic mutating algorithms must be extensions of `MutableCollection where Element: Movable`, for which a trait alias might again be useful, but I wouldn't rush to create it.

**Comment by kyouko-taiga** [comment]
> I guess remote things are immutable, right?

I think you meant immovable, right?

That is a good question, honestly. They are not movable to storage with a longer lifetime, but I don't see an obvious reason why you wouldn't be able to move remote things around if they stay in the same array. 🤷‍♀️ 

> it's hard to imagine what kinds of things should be immovable

Yep. I _think_ locks are immovable objects? I can also think about an object that would expose its address as a salient feature (e.g., its identity).

```swift
type A: Comparable, Deinitializable {
  public memberwise init
  public fun infix== (_ other: Self) -> Bool {
    pointer[to: self] == pointer[to: other]
  }
  public fun infix< (_ other: Self) -> Bool {
    pointer[to: self] < pointer[to: other]
  }
}
```

**Comment by dabrahams** [comment]
> I think you meant immovable, right?

Yeah, that's what I meant.

> They are definitely not movable to storage with a longer lifetime, but we could safely move remote things around if they stayed in the same array. 

Would you even be able to track which remote things are safe to exchange?  They won't all have the same lifetime constraints.

> At the same time I really don't think we want to be able to reassign a remote part to a different part, that is &x = y should mean "move y to the storage referred by x" and not "make x refer to the storage of y".

I don't think it should mean anything about storage; that's a bit too low-level.  I think it should mean “bind x to an object having the value of y.”  In which case, I *think* there's no problem.  But maybe I'm missing something.

**Comment by kyouko-taiga** [comment]
> Would you even be able to track which remote things are safe to exchange?

I think it wouldn't matter, because the collection would be "anchored" to the innermost lifetime it may contain.

To illustrate, assume remote parts are pointers in C++. The pictures would look like this:
```c++
bool leq(int* a, int* b) { return *a < *b; }

void g(int*m, int *n) {
  std::vector v = {m, &n};
  std::sort(v.begin(), v.end(), &leq);
}

void f(int* n) {
  int y = 1;
  g(n, &y);
}

int main() {
  int x = 2;
  f(&x);
}
```
In this example, `v` is anchored to the scope of `f` because the innermost "remote part" that it contains is defined in that scope. It is safe to swap pointers because none of them live shorter than that. That means `std::sort` cannot violate temporal safety when it accesses these remote parts.

This example is using lifetime reasoning, though no annotation is needed because the scope discipline let us guarantee that everything is safe. An annotation would be needed iff `v` was allowed to escape `g`. So I believe there's a way to implement a weaker form of Rust-like borrowed references in the form of remote types as long as we enforce scope discipline, which pairs relatively well with subscripts.

Note that seeing at remote parts as pointers make sense if we think about Rust references, because Rust references are "reassignable". You can make a `&mut T` refer to a different storage after it's been initialized. That's because there's a first-class distinction between `&mut T` and `T`. So really, AFAICT `&mut T` in Rust works like `T*` in C++.

Now one question is: should remote part behave more like Rust borrowed references (i.e., C++ pointers) or more like C++ references? I would lean toward the latter. If we go this road, then I guess `remote inout T` is not `Movable`, regardless of `T`'s own conformance to the same trait. In fact, I think `remote inout T` would never conform to any trait except `Deinitializable`. But sadly that approach has other caveats.

**Comment by dabrahams** [comment]
>Note that seeing remote parts as pointers makes sense if we think about Rust references, 

Sure, but Hylo's raison d'être  is to present a mental model where you _don't_ think in terms of references and pointers.  That's why I don't want to describe the meaning of `=` in terms of storage, and why this question doesn't seem right to me either:

> should remote part behave more like Rust borrowed references (i.e., C++ pointers) or more like C++ references?

Remote parts are simply parts that are borrowed values, and should behave as such to a maximal extent.  That's why we named them that way.  Think of them as semantically having been copied (`let`) or temporarily moved (`inout`) out of their owning context, without any actual move or copy requirement, and with a different storage layout.  Otherwise, operations on them should function as though they were ordinary stored parts.  I believe the "temporarily moved" tells us everything we need to know about assignment: whatever value is in a given remote part must, at the end of the part's lifetime, be semantically “returned” (even though it's already in the right storage) to the original source of that part.  Therefore, an assignment into a remote part must effectively assign the underlying storage.

**Comment by kyouko-taiga** [comment]
> I believe the "temporarily moved" tells us everything we need to know about assignment

That's a very nice way to frame it.

It occurred to me that my main issue can be boiled down to thinking about the semantics of `Movable` w.r.t. the storage of an instance of `Self`. For example, `T::Movable.take_value(from:).set` is supposed to initialize storage of type `T` by consuming a value of the same type. If `T` is replaced by `remote inout Int`, then it stands to reason that we'd be talking about the assignment of the value of a pointer to an integer.

But, thinking of remote types as type layout + some rights/duty suggests that the meaning of `Self` for `remote inout T` is just `T`. That's a huge step forward in the understanding of remote types, at least for me.

It might also be possible to unify `indirect` and `remote` under this framework because I believe `indirect T` could be synonym for `remote sink T` (note: I got somewhat convinced that `remote sink T` != `T`). One annoying question is about the lifetime constraint imposed by the storage where `T` is actually stored. One way to think of it may be that `indirect T` is `remote sink T` where the storage never goes out of scope; it just becomes inaccessible at some point.

**Comment by dabrahams** [comment]
> But, thinking of remote types as type layout + some rights/duty suggests that the meaning of `Self` for `remote inout T` is just `T`.

Yeah, that's how I think of it.  I thought that was part of what it meant when we said `remote inout T` is not a first-class type?

> It might also be possible to unify indirect and remote under this framework because I believe `indirect T` could be synonym for `remote sink T` 

Fascinating idea!  If `indirect` means what it means in Swift, an `indirect T` has some destruction semantics that we need to account for (decrementing a refcount and/or destroying/deallocating the `T`).  Does that fit with the way you'd understand `remote sink T`?

> the storage never goes out of scope

Sorry, I don't know what it means for storage to be in or out of scope.  The way I understand scopes, they are about name visibility.  Could you explain?

**Comment by kyouko-taiga** [comment]
> If `indirect` means what it means in Swift, an `indirect T` has some destruction semantics that we need to account for [...] Does that fit with the way you'd understand `remote sink T`?

I believe it does. If a part has type `remote sink T`, then its whole has a responsibility to take care of it, via deinitialization or other means. The original owner of that part will have relinquished this responsibility.

For example:

```
fun main() {
  let p = (x: 1, y: 1)
  let q = (x: remote sink p.x, 2)
  // At this point, `p` no longer has the responsibility to deinitialize `p.x` because `q` does.
  print(p.y)
  // `p` is no longer in use; `p.y` may  be deinitialized from here.
  print(q)
  // `q` is no longer in use; it may be deinitialized from here, thereby also deinitializing `p.x`.
}
```

When we initialized `q`, `p` relinquished ownership of `p.x` and with it the responsibility to deinitialize this particular part. This responsibility has been picked up by `q`. What's interesting is that the storage containing `p`'s value (which is no partially deinitialized) must exist longer than `p`'s useful lifetime. So in a sense `p` is still responsible to maintain storage for its remote parts, no matter the capability they have (`let`, `inout`, ...).

> Sorry, I don't know what it means for storage to be in or out of scope. The way I understand scopes, they are about name visibility. Could you explain?

Perhaps I should have talked about stack frame rather than scope. The issue is about guaranteeing that the storage containing a remote part doesn't get deallocated (or reused) while that remote part may be accessed.

In the example above, the fact that `q.x` is remote means that `q` relies on the storage holding that value to exist at least as long as it is alive. So we are not allowed to deallocate the storage of `p`'s value before the last use of `q`. In fact, we're not even allowed to touch `p.x` before the last use of `q`.

Normally, objects can freely move from one binding to another as long as they have no projected parts in use. But when remote parts are involved, objects can no longer move to a binding that is accessible beyond the lifetime of the storage holding these remote parts, even when they are "remote".

`indirect` is special in the sense that the storage holding the remote part can be kept alive as long as we need. It is not constrained to the stack as in the above example, where `p` is a stack allocation that will be deallocated when `main` returns. For all intents and purposes, the storage can be thought as having infinite lifetime. Of course we want to deallocate `indirect` storage _eventually_ but I think that's easy to do. (Happy to expand on the details if you want.)

The key observation is that storage lifetime and object lifetime are separate concepts. We can make this observation with a simpler example:

```
public fun main() {
  var x: Int
  &x = 1 // lifetime of "1" starts and ends here "1"
  &x = 2 // lifetime of "2" starts and ends here "2"
}
```
